### ISSUE TITLE
OData 8.x with ASP.NET 5 always returns 401 Unauthorized on api endpoints

### ISSUE THREAD

**Issue by JayCallas** [OData/AspNetCoreOData#289]
New ASP.NET 5 api project that is trying to use Bearer token authentication against OData 8 endpoints. All requests result in 401 Unathoized being returned as if the token is not even being checked.

The same code works fine with ASP.NET 5 and OData 7.x.

```
public void ConfigureServices(IServiceCollection services) {
	...
	if (Environment.IsLocal()) {
		services
			.AddAuthentication("BasicAuthentication")
			// MockAuthenticatedUser inherits from AuthenticationHandler for local development/testing purposes
			.AddScheme<AuthenticationSchemeOptions, MockAuthenticatedUser>("BasicAuthentication", null);
	}
	else {
		services.AddDistributedMemoryCache();

		services
			.AddAuthentication(IdentityServerAuthenticationDefaults.AuthenticationScheme)
			.AddIdentityServerAuthentication(options => {
				options.Authority = ssoSettings.BaseUri.ToString();
				options.ApiName = "xyz";
				options.ApiSecret = ssoSettings.ClientSecret;

				options.EnableCaching = true;
			});
	}
	...
	services.AddSwaggerGen(c => {
		...
	});
	...
	services
		.AddControllers(options => {
			var policy = new AuthorizationPolicyBuilder().RequireAuthenticatedUser().Build();
			options.Filters.Add(new AuthorizeFilter(policy));
		})
		.AddOData(opt => {
			opt
				.Count()
				...
				.AddRouteComponents("api", EdmModelBuilder.GetEdmModel());
		});
}

public void Configure(IApplicationBuilder app) {
	...
	app.UseAuthorization();
	app.UseAuthentication();

	app.UseEndpoints(endpoints => {
		endpoints
			.MapControllers()
			.RequireAuthorization();
	});
}
```

The issue does not seem to be specific to `AddIdentityServerAuthentication()`. During local development, using my `MockAuthenticatedUser` handler, the same 401 Unauthorized is returned. Breakpoints in the handler will trigger when Swagger page is viewed -- indicating that authentication is configured correctly -- but does **not** trigger when api requests are made.

Unable to find documentation or samples in this repo on proper configuration.

**Comment by xuzhg** [comment]
@JayCallas Would you please share a repro for me to debug?

**Comment by JayCallas** [comment]
@xuzhg -- I will see if I can strip down my code to just the minimum but it is pretty much what you see in the OP. In the meantime, is there an example or documentation I can refer to for adding authentication to OData 8.0 project?

**Comment by JayCallas** [comment]
I was able to create a simple repo. Comments about behavior can be found in `Startup.cs`.

[Announcements-API.zip](https://github.com/OData/AspNetCoreOData/files/7080066/Announcements-API.zip)

**Comment by xuzhg** [comment]
@JayCallas We haven't had example to add authentication to OData 8.0 project. But, I do want to create one.

For your sample, if i move your swagger middleware after authentication, it can hit the breaking:

```C#
                       app.UseRouting();

			app.UseCors();

			app.UseAuthorization();
			app.UseAuthentication();

			app.UseSwagger();
			app.UseSwaggerUI(c => c.SwaggerEndpoint("/swagger/v1/swagger.json", "Announcements API v1"));

			app.UseEndpoints(endpoints => {
				endpoints
					.MapControllers()
					.RequireAuthorization();
			});

```

Here's my snapshot:

![image](https://user-images.githubusercontent.com/9426627/131420941-e9594d85-323b-45e8-bfde-cdb6008177ca.png)

**Comment by JayCallas** [comment]
@xuzhg I am getting the exact opposite behavior. When I moved the Swagger lines to what you show in your screenshot, the Swagger endpoint starts returning 401 Unauthorized. The only call that is triggering the breakpoint is the controller endpoint that is set to anonymous.

![image](https://user-images.githubusercontent.com/15733663/131504137-2b9980e9-5087-41e9-ad77-0ca90263e940.png)

![image](https://user-images.githubusercontent.com/15733663/131504203-55cc7135-3164-4ef8-80ac-a59110a101cb.png)

**Comment by xuzhg** [comment]
@JayCallas Is '401 Unauthorized' not the expected behavior? 
Anyway, do you think it's related to OData functionality.

**Comment by JayCallas** [comment]
@xuzhg Yes, it **only started happening after I added OData** to my service. When I had **non-OData** controllers, authentication worked as expected.
